src/vmm/src/pipe.rs

```diff
@@ -0,0 +1,69 @@
+// Copyright 2020 Sartura All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+#![cfg(target_arch = "x86_64")]
+
+use std::io::{Read, Write};
+
+use std::fs::{File};
+use std::os::unix::io::{AsRawFd, RawFd};
+
+use nix::sys::stat;
+use nix::unistd;
+use std::path::PathBuf;
+
+/// Newtype for piping the output of guest cmd.
+pub(crate) struct PipeReaderWrapper(PathBuf, File);
+
+impl PipeReaderWrapper {
+    pub fn new(buf: PathBuf) -> PipeReaderWrapper {
+        match unistd::mkfifo(&buf, stat::Mode::S_IRWXO) {
+            Ok(_) => (),
+            Err(err) => panic!("Error creating fifo: {}", err),
+        }
+        let pipe = match File::open(&buf) {
```

Comment:
I'm thinking of this a bit differently: what if you created the `Read`er somewhere else, and worked with that instead, passing it on to different implementors? That way you could have a generic serial reader that only cares about the sink being `Read` and `AsRawFd`, and 2+ different instantiations:
1. `io::Stdin`, which conveniently is both `Read` and `AsRawFd` already
2. a pipe
3. (why not?) the unix domain socket suggested by @sameo 

This is an idea somewhat orthogonal to the changes in this PR - the rest of my comments directly address the individual changes, but I'd like to get your opinion on this approach.